agent/app/agent.go

```diff
@@ -377,6 +396,44 @@ func (agent *ecsAgent) doStart(containerChangeEventStream *eventstream.EventStre
 		deregisterInstanceEventStream, client, state, taskHandler)
 }
 
+// waitUntilInstanceInService Polls IMDS until the target lifecycle state indicates that the instance is going in
+// service. This is to avoid instances going to a warm pool being registered as container instances with the cluster
+func (agent *ecsAgent) waitUntilInstanceInService() error {
+	var err error
+	var targetState string
+	// Poll while the instance is in a warmed state or while waiting for the data to be populated.
+	// If the data is not populated after a certain number of polls, then stop polling and return the not found error.
+	// The polling maximum does not apply to instances in the warmed states
+	for i := 0; i < asgLifeCyclePollMax || targetState != ""; i++ {
+		targetState, err = agent.getTargetLifeCycle()
+		// stop polling if the retrieved state is in service or we get an unexpected error
+		if (targetState == inServiceState) || (err != nil && !strings.Contains(err.Error(), "status code: 404")) {
+			break
+		}
+		time.Sleep(asgLifeCyclePollWait)
+	}
+	return err
+}
+
+// getTargetLifecycle obtains the target lifecycle state for the instance from IMDS. This is populated for instances
+// associated with an ASG
+func (agent *ecsAgent) getTargetLifeCycle() (string, error) {
+	var targetState string
+	var err error
+	backoff := retry.NewExponentialBackoff(targetLifecycleBackoffMin, targetLifecycleBackoffMax, targetLifecycleBackoffJitter, targetLifecycleBackoffMultiple)
+	for i := 0; i < targetLifecycleMaxRetryCount; i++ {
+		targetState, err = agent.ec2MetadataClient.TargetLifecycleState()
+		if err == nil {
+			break
+		} else if i < targetLifecycleMaxRetryCount {
+			time.Sleep(backoff.Duration())
+		}
+		seelog.Warnf("Error when getting intended lifecycle state: %v", err)
```

Comment:
can we print this error before sleeping?

Reply:
Should it be a warn or a debug?

Reply:
Good point. Was initially thinking warn so we are aware of errors that occur, which aren't always expected, though if it results in a failure then we are logging it before exiting anyway. I'll change to debug